### PR TITLE
Refactor XmlQualifiedName

### PR DESCRIPTION
**Pull request by kronic**

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 1]
```diff
@@ -36,83 +34,69 @@ public class XmlQualifiedName
...
         public static bool operator ==(XmlQualifiedName? a, XmlQualifiedName? b)
         {
-            if ((object?)a == (object?)b)
+            if (ReferenceEquals(a, b))
+            {
                 return true;
+            }
 
-            if (a is null || b is null)
+            if (ReferenceEquals(a, null) || ReferenceEquals(b, null))
```
Comment (drieseng) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:96]: use `is null`?

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:96]: https://stackoverflow.com/questions/12465961/is-referenceequalsmyobject-null-better-practice-than-myobject-null

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:96]: `is null` is different from `== null`.  The former is a pure null check, the latter may invoke any operator== overloads on the type.  The use of `is null` was fine here and needn't be replaced by ReferenceEquals.

**Comment by kronic** [comment]
@drieseng @stephentoub done

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 1 above)
Comment (drieseng) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs]: thanks for the confirmation.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 2]
```diff
@@ -130,42 +114,41 @@ public override bool Equals([NotNullWhen(true)] object? other)
...
+                XmlConvert.ToUri(Namespace);
             }
         }
 
         internal void Atomize(XmlNameTable nameTable)
         {
-            Debug.Assert(_name != null);
-            _name = nameTable.Add(_name);
-            _ns = nameTable.Add(_ns);
+            Debug.Assert(Name != null);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:144]: considering we have nullable annotations enabled, Name is non-null and compiler doesn't produce any warnings consider removing the assert

**Review by krwq — APPROVED** [review]

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 3]
```diff
@@ -130,42 +114,41 @@ public override bool Equals([NotNullWhen(true)] object? other)
         /// </devdoc>
         public static string ToString(string name, string ns)
```
Comment (kronic) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:115]: @krwq Should I also replace with
public static string ToString(string? name, string? ns)

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 4]
```diff
@@ -36,83 +34,69 @@ public class XmlQualifiedName
...
 
         /// <devdoc>
         ///    <para>[To be supplied.]</para>
         /// </devdoc>
-        public string Namespace
-        {
-            get { return _ns; }
-        }
-
+        public string Namespace { get; private set; }
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:44]:
```suggestion
        public string Namespace { get; }
```

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 5]
```diff
@@ -36,83 +34,69 @@ public class XmlQualifiedName
...
         /// <devdoc>
         ///    <para>[To be supplied.]</para>
         /// </devdoc>
-        public string Name
-        {
-            get { return _name; }
-        }
-
+        public string Name { get; private set; }
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:48]:
```suggestion
        public string Name { get; }
```

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:115]: This does not expect a null name.  Looks like it should be `string? ns`, though, and the ref should be updated for that as well.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:48]: Ah, nevermind, I see we set it later outside of the constructor.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 4 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:44]: see  internal void Init(string? name, string? ns)

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 5 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:48]: see internal void Init(string? name, string? ns)

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 3 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:115]: How can I ref update?

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:115]: Im run ./dotnet build --no-incremental /t:GenerateReferenceSource
Get error
 D:\GitHub\runtime\src\coreclr\tools\dotnet-pgo\dotnet-pgo-pack.proj : error MSB4057: в проекте нет целевого объекта "Clean".

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs]: Just add the `?` to
https://github.com/dotnet/runtime/blob/f169b71c5403a1fc0523c93f3df64be3e109b34e/src/libraries/System.Xml.ReaderWriter/ref/System.Xml.ReaderWriter.cs#L768

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs]: @stephentoub thanks, done

**Comment by kronic** [comment]
@krwq, @stephentoub done

**Review by krwq — APPROVED** [review]

**Comment by krwq** [comment]
Thanks @kronic!

**Pull request merged**